pkg/intoto/metadata.go

```diff
@@ -9,89 +9,148 @@ import (
 	}
 
-	links := make(map[string][]byte)
-	files, err := ioutil.ReadDir(linkDir)
+	metadata := Metadata{Data: data, PublicKeys: filenames}
+	custom := Custom{InToto: metadata}
+	filename = gun + "/in-toto-metadata/" + filename
+	rootLayout.Filename = filename
+	rootLayout.Custom = custom
+	return rootLayout, nil
+}
+
+// GetLinks reads link metadata off disk into a Links struct
+func GetLinks(gun string, dir string) (Links, error) {
+	fileinfos, err := ioutil.ReadDir(dir)
 	if err != nil {
-		return nil, fmt.Errorf("cannot read links directory %v: %v", linkDir, err)
+		return nil, fmt.Errorf("cannot read links directory %v: %v", dir, err)
 	}
-	for _, f := range files {
-		// TODO - Radu M
-		//
-		// robust check if file is actually a link
-		if !strings.Contains(f.Name(), ".link") {
-			continue
+
+	// NOTE: We assume that the base directory used to hold the links are unique
+	// (e.g., identified by the digest of the first link metadata file corresponding to the first step).
+	// This is so that different versions of links corresponding to the same root layout can be safely isolated from each other.
+	// TODO: If they are not unique, should we raise an error later when adding them to the TUF targets metadata?
+	digest := filepath.Base(dir)
+	links := make(Links)
+	for _, fileinfo := range fileinfos {
+		filename := fileinfo.Name()
+		if !strings.Contains(filename, ".link") {
+			return nil, fmt.Errorf("%s does not have a .link suffix", filename)
 		}
-		b, err := ioutil.ReadFile(filepath.Join(linkDir, f.Name()))
+
+		data, err := readFile(filepath.Join(dir, filename))
 		if err != nil {
-			return nil, fmt.Errorf("cannot get canonical JSON from file %v: %v", f.Name(), err)
+			return nil, err
 		}
-		links[f.Name()] = b
+
+		metadata := Metadata{Data: data}
+		custom := Custom{InToto: metadata}
+		filename = gun + "/in-toto-metadata/" + digest + "/" + filename
+		links[filename] = custom
 	}
 
-	m := &Metadata{
-		Key:    k,
-		Layout: l,
-		Links:  links,
+	return links, nil
+}
+
+// GetBundleCustom returns a Custom struct pointing to a list of root layout and links
+func GetBundleCustom(rootLayout RootLayout, links Links) Custom {
+	filenames := make([]string, len(links)+1)
+	filenames = append(filenames, rootLayout.Filename)
+	for filename := range links {
+		filenames = append(filenames, filename)
 	}
+	metadata := Metadata{Links: filenames}
+	return Custom{InToto: metadata}
+}
+
+// WriteMetadataFiles writes the content of a metadata object into files in a directory
+// TODO
+func WriteMetadataFiles(m *Metadata, dir string) error {
+	return fmt.Errorf("not implemented")
+}
 
-	raw, err := canonicaljson.MarshalCanonical(m)
+func readFile(filename string) ([]byte, error) {
+	bytes, err := ioutil.ReadFile(filename)
 	if err != nil {
-		return nil, fmt.Errorf("cannot encode in-toto metadata into canonical json %v: %v", m, err)
+		return nil, err
 	}
-
-	return canonicaljson.RawMessage(raw), nil
+	return bytes, nil
```

Comment:
So this function returns `ioutil.ReadFile`, right?

Reply:
💯 yes, please go for it, sorry missed it after refactoring it